File: factorPowers.go
/*
 =============================================
 Author:      Aloyts
 Create date: 2018 August 06
 Description: Factoring powers of 2
 =============================================
*/
package main

import ("fmt")
import ("math")

func main() {
	var factors [50]float64
	
	var totalVal float64 = 17
	var powCount = totalVal
	
	var maxPower float64
	var rem float64
	
	//var i float64
	var j int = 0
	
	maxPower = math.Floor(math.Log2(totalVal))
	rem = totalVal - math.Pow(2, maxPower)
	fmt.Println(maxPower)
	fmt.Println(rem)
	
	if(rem!=0){
		for i:=maxPower; i >= 0 ; i-- {
			if(powCount - math.Pow(2,i) >= 0){
          			factors[j] = math.Pow(2,i)
			        powCount -= math.Pow(2,i)
				j++
				fmt.Printf("nextPow=%.0f\n", math.Pow(2,i))
				
				fmt.Println("j=", j)
        		}
		}
	}

}

File: factorPowers.dart
/*
 =============================================
 Author:	  Aloyts
 Create date: 2018 August 16
 Description: Factoring powers of 2
 =============================================
*/

import 'dart:math';

void main() {
  
  print(factorPowers(15));
}

String factorPowers(powInt){

  int maxPower = (log(powInt)/log(2)).floor();
  
  print("maxPower=" + maxPower.toString());
  
  int rem = (powInt - pow(2, maxPower));
	
  print("Remainder=" + rem.toString());
  
  String factors = "";
  
  if(rem!=0)
  {
    
    int powCount = powInt;

      for(int i=maxPower;i>=0;i--)
      {

          if(powCount - pow(2,i) >= 0)
          {
            factors += pow(2,i).toString() + ",";
            powCount -= pow(2,i);
          }
      }
    factors = factors.substring(0, factors.length-1);  //remove trailing comma.
    
}

return factors;
}